### PR TITLE
Merge sort 3 way in c language

### LINKED ISSUE #521: 3 Way Merge Sort
3 Way merge sort designed in c language

### PR DESCRIPTION
**Pull request by KavithaVarughese**
Fixes #521

#### Checklist:
- [x] space indentation.
- [x] Coding conventions are followed.
- [ ] Sample Input / Output.
- [ ] Logic Documentation (Comments).
- [x] File names are correct.

#### Changes proposed in this pull request:
- Added a C program for 3 Way Merge Sort

#### Languages Used:
- C 

#### Files Added:
- Merge_Sort_3_Way.c

### CODE REVIEW

File: Merge_Sort_3_Way/Merge_Sort_3_Way.c [hunk 1]
```diff
@@ -0,0 +1,123 @@
+#include<stdio.h>
+//#include<stdlib.h>
```
Comment (marwahmanbir) [inline Merge_Sort_3_Way/Merge_Sort_3_Way.c]: Hey! Please remove this line if the commented header file is for no use.

File: Merge_Sort_3_Way/Merge_Sort_3_Way.c [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+void mergesort(long int *arr1, int p , int q ,long int *arr2);
+int merge(long int *arr1,int l,int m1, int m2, int h, long int *arr2 );
+
+int main()
+{ 
+
```
Comment (marwahmanbir) [inline Merge_Sort_3_Way/Merge_Sort_3_Way.c]: Please remove unnecessary blank lines like these. Blank lines should only be added
- Between sections of a source file
- Between class and interface definitions
- Between methods
- Between the local variables in a method and its first statement
- Before a block or single-line comment
- Or between logical sections inside a method to improve readability

File: Merge_Sort_3_Way/Merge_Sort_3_Way.c [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+  scanf("%ld",&n);
+  
+  long int* arr1 = (long int *)malloc(n * sizeof(long int));
+  long int* arr2=(long int *)malloc (n * sizeof(long int ));
+  
+  int i=0;
+  long int num;
+
+  
+
```
Comment (marwahmanbir) [inline Merge_Sort_3_Way/Merge_Sort_3_Way.c]: Too many unnecessary blank lines.

File: Merge_Sort_3_Way/Merge_Sort_3_Way.c [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+  while(i<n)
+  {
+    printf("Enter the %d number : ", i + 1);
+    scanf("%ld",&num);
+	  arr1[i]=num;
+	  arr2[i]=num;
+	  i++;
```
Comment (marwahmanbir) [inline Merge_Sort_3_Way/Merge_Sort_3_Way.c]: Fix indentation - 4 spaces in place of tab, 8 spaces in place of 2 tabs, and so on...

File: Merge_Sort_3_Way/Merge_Sort_3_Way.c [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+   int j=m1;
+   int p=m2;
+   //int q=h;
+   int k =l;
+   while((i<m1) && (j< m2) && ( p< h))
+   {
+	if (arr2[i]<arr2[j])
+	{
+		if(arr2[i]< arr2[p])
+			arr[k++] = arr2[i++];
```
Comment (marwahmanbir) [inline Merge_Sort_3_Way/Merge_Sort_3_Way.c]: Fix indentation.

File: Merge_Sort_3_Way/Merge_Sort_3_Way.c [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+   	arr[k++] = arr2[i++];
+   while(j<m2)
+   	arr[k++] = arr2[j++];
+   while(p<h)
+   	arr[k++] = arr2[p++]; 
+  
+   
+}
+
+
```
Comment (marwahmanbir) [inline Merge_Sort_3_Way/Merge_Sort_3_Way.c]: Remove extra lines. Add sample input-output at the end of file, then add a single blank line at the end of file.

**Comment by marwahmanbir** [comment]
Add comments to explain the logic behind your code so that it is easily understandable by anyone who reads it. Fix checklist above (remove extra bullets).
Consider the comments added in the review for whole code and not only for the lines they're mentioned against.

File: Merge_Sort_3_Way/Merge_Sort_3_Way.c [hunk 7]
```diff
@@ -0,0 +1,123 @@
+#include<stdio.h>
+//#include<stdlib.h>
+
+void mergesort(long int *arr1, int p , int q ,long int *arr2);
+int merge(long int *arr1,int l,int m1, int m2, int h, long int *arr2 );
```
Comment (SakshiShreya) [inline Merge_Sort_3_Way/Merge_Sort_3_Way.c]: There should be no space before comma and one space after each comma.

File: Merge_Sort_3_Way/Merge_Sort_3_Way.c [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+int main()
+{ 
+
+  int length;
+
+  long int n;
+  printf("Number of elements in the array : ");
+  scanf("%ld",&n);
```
Comment (SakshiShreya) [inline Merge_Sort_3_Way/Merge_Sort_3_Way.c]: Add a space after comma.

File: Merge_Sort_3_Way/Merge_Sort_3_Way.c [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+  
+  long int* arr1 = (long int *)malloc(n * sizeof(long int));
+  long int* arr2=(long int *)malloc (n * sizeof(long int ));
+  
+  int i=0;
```
Comment (SakshiShreya) [inline Merge_Sort_3_Way/Merge_Sort_3_Way.c]: Add space before and after the equals sign.

File: Merge_Sort_3_Way/Merge_Sort_3_Way.c [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+{ 
+
+  int length;
+
+  long int n;
+  printf("Number of elements in the array : ");
+  scanf("%ld",&n);
+  
+  long int* arr1 = (long int *)malloc(n * sizeof(long int));
+  long int* arr2=(long int *)malloc (n * sizeof(long int ));
```
Comment (SakshiShreya) [inline Merge_Sort_3_Way/Merge_Sort_3_Way.c]: - Add space before and after equals sign.
- Remove space after `malloc`
- Remove space after the last `long int`

File: Merge_Sort_3_Way/Merge_Sort_3_Way.c (hunk 6 above)
Comment (SakshiShreya) [inline Merge_Sort_3_Way/Merge_Sort_3_Way.c]: - Add space before and after all the operators.
- Add space after commas and remove the spaces before them.

**Comment by SakshiShreya** [comment]
Fix the PR template

**Comment by jainaman224** [comment]
@KavithaVarughese Are you working on it?

**Comment by jainaman224** [comment]
Hey,
Closing this for now. Whenever you are working on it. Please reopen it.

**Pull request closed without merging**